File: src/agent_search_1688/tools/registry.py
"""1688 Agent Search 的受控工具注册表。"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Callable


ToolHandler = Callable[[dict[str, Any]], dict[str, Any]]


@dataclass(frozen=True)
class ToolEntry:
    name: str
    description: str
    input_schema: dict[str, Any]
    handler: ToolHandler
    parallel_safe: bool = False

    def as_mcp_definition(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "description": self.description,
            "inputSchema": self.input_schema,
        }


class ToolRegistry:
    def __init__(self) -> None:
        self._entries: dict[str, ToolEntry] = {}

    def register(self, entry: ToolEntry) -> None:
        if entry.name in self._entries:
            raise ValueError(f"工具名称重复：{entry.name}")
        self._entries[entry.name] = entry

    def definitions(self) -> list[dict[str, Any]]:
        return [entry.as_mcp_definition() for entry in self._entries.values()]

    def dispatch(self, name: str, arguments: dict[str, Any]) -> dict[str, Any]:
        entry = self._entries.get(name)
        if entry is None:
            raise KeyError(f"未注册工具：{name}")
        return entry.handler(arguments)

    def is_parallel_safe(self, name: str) -> bool:
        entry = self._entries.get(name)
        return bool(entry and entry.parallel_safe)

File: src/tools/registry.ts
export type JsonSchema = Record<string, unknown>;
export type ToolResult = Record<string, unknown>;

export interface ToolEntry {
  readonly name: string;
  readonly description: string;
  readonly inputSchema: JsonSchema;
  readonly handler: (arguments_: Record<string, unknown>) => ToolResult | Promise<ToolResult>;
  readonly parallelSafe?: boolean;
}

export interface McpToolDefinition {
  readonly name: string;
  readonly description: string;
  readonly inputSchema: JsonSchema;
}

export class ToolRegistry {
  readonly #entries = new Map<string, ToolEntry>();
  readonly #closeHandlers: Array<() => void | Promise<void>> = [];
  #closed = false;

  register(entry: ToolEntry): void {
    if (this.#entries.has(entry.name)) throw new Error(`工具名称重复：${entry.name}`);
    this.#entries.set(entry.name, entry);
  }

  addCloseHandler(handler: () => void | Promise<void>): void {
    this.#closeHandlers.push(handler);
  }

  definitions(): McpToolDefinition[] {
    return [...this.#entries.values()].map((entry) => ({
      name: entry.name,
      description: entry.description,
      inputSchema: entry.inputSchema,
    }));
  }

  async dispatch(name: string, arguments_: Record<string, unknown>): Promise<ToolResult> {
    if (this.#closed) throw new Error("工具注册表已关闭");
    const entry = this.#entries.get(name);
    if (!entry) throw new Error(`未注册工具：${name}`);
    return await entry.handler(arguments_);
  }

  isParallelSafe(name: string): boolean {
    return this.#entries.get(name)?.parallelSafe === true;
  }

  async close(): Promise<void> {
    if (this.#closed) return;
    this.#closed = true;
    for (const handler of this.#closeHandlers.reverse()) await handler();
  }
}
